archive/unix_pack.ksy

```diff
@@ -0,0 +1,107 @@
+meta:
+  id: unix_pack
+  title: Unix pack(1)-compressed data
+  application:
+    - pack(1)
+    - unpack(1)
+    - pcat(1)
+  # The extension is specifically lowercase.
+  # Uppercase .Z is used for compress(1)-compressed data.
+  file-extension: z
+  xref:
+    justsolve: Pack
+    wikidata: Q11239298
+  license: CC0-1.0
+  endian: be
+doc: |
+  The Unix pack compression format, a very basic compression format based on
+  Huffman coding. The Unix `pack(1)` utility could compress files into
+  this format, and could be decompressed using the utilities `unpack(1)`
+  and `pcat(1)`.
+  
+  The pack format is long obsolete - it was superseded in the 1980s
+  by the LZW-based compression format of `compress(1)` (which is itself
+  almost never used anymore on modern Unix derivatives). Despite this,
+  the `gzip(1)` utility still supports decompressing (but not compressing)
+  data in the pack compression format.
+doc-ref:
+  - 'https://www.vidarholen.net/contents/blog/?p=691'
+  - 'https://github.com/koalaman/pack/blob/master/pack.hs'
+  - 'https://git.savannah.gnu.org/cgit/gzip.git/tree/unpack.c'
+  # The source code for the original Unix pack and unpack utilities
+  # (as of Research Unix, Edition 8) can be found in the archive v8.tar.bz2
+  # in the directory usr/src/cmd/pack.
+  - 'https://www.tuhs.org/Archive/Distributions/Research/Dan_Cross_v8/'
+seq:
+  - id: magic
+    contents: [0x1f, 0x1e]
+  - id: len_uncompressed
+    type: u4
+    doc: The length of the data when uncompressed, in bytes.
+  - id: tree
+    type: tree
+    doc: The Huffman tree used to compress the data.
+  - id: data
+    size-eos: true
+    doc: |
+      The compressed data. This is a sequence of Huffman codes (variable-length
+      bit sequences) stored in most-significant bit first order without any
+      padding/alignment between individual codes. If the coded data does not
+      end on a byte boundary, it is padded with zeroes at the end (i. e. in the
+      least significant bits).
+types:
+  tree_level:
+    params:
+      - id: level_number
+        type: u1
+        doc: This level's number.
+    seq:
+      - id: leaves
+        type: u1
+        repeat: expr
+        # See doc of tree.level_leaf_counts for an explanation of
```

Comment:
Why not to just move the explaination into the `doc` of this field?
Also it may be beneficial to move the calculation into an instance.

Reply:
I'm putting this in a YAML comment so that it only appears in the spec itself, and not in the code generated by ksc. I think the explanation is only relevant to people reading the source code of the spec, and not to people using the generated code (they will only access the `leaves` list itself and won't care about the internal calculations).

Good idea to move this into an instance though, will do that.

I don't think it makes sense to put this into the `doc`, because it's not relevant to users of the spec - they will only use the `leaves` list and won't have to deal with internal calculations like this. By putting the explanation in a YAML comment, it will be visible to people reading the spec, but it will be discarded when the spec is compiled.